### PR TITLE
STM32F7 stuff

### PR DESCRIPTION
**Pull request by H2OBrain**
Changes in order
- enable exti headers (from f4)
- enable fsmc (from f4)
- add LTDC_BASE definition to the f7 memorymap.h to be compatible with f4 (for ltdc)
- enable ltdc (from f4)
- extend f7 memorymap.h and rcc.h for dsi
- enable dsi headers (from f4)
- fix typos in dsi headers
- move static inline functions to c
- 'synchronize' rcc.h with f4
- partially combine rcc.c with f4
- enable dma2d headers (from f4)
- fix invalid color definition in dma2d header
- extend/reorganise dsi header
    - backwards compatible
    - documented fields
    - bitfields grouped in typedef enum { .. } dsi_regname_t;
    - field value definitions also grouped in typedef enum { .. } dsi_regname_fieldname_t;
- extended SCB_TypeDef (core_cm3.h)
- corrected comment in F7 Makefile (some MCUs do have double-precision fpu)

### CODE REVIEW

File: include/libopencm3/stm32/f7/exti.h [hunk 1]
```diff
@@ -0,0 +1,42 @@
+/** @defgroup exti_defines EXTI Defines
+ *
+ * @brief <b>Defined Constants and Types for the STM32F4xx External Interrupts
```
Comment (karlp) [inline include/libopencm3/stm32/f7/exti.h]: typo

File: include/libopencm3/stm32/f7/exti.h [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+ * @version 1.0.0
+ *
+ * @date 14 January 2014
+ *
+ * LGPL License Terms @ref lgpl_license
+ */
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * Copyright (C) 2013 Piotr Esden-Tempski <piotr@esden.net>
```
Comment (karlp) [inline include/libopencm3/stm32/f7/exti.h]: I really don't think this is the right copyright statement for two includes.

File: lib/stm32/f7/Makefile [hunk 3]
```diff
@@ -47,8 +47,10 @@ OBJS		+= crc_common_all.o crc_v2.o
 OBJS		+= dac_common_all.o
 OBJS		+= desig.o
 OBJS		+= dma_common_f24.o
+OBJS		+= exti_common_all.o
 OBJS		+= flash_common_all.o flash_common_f.o flash_common_f24.o flash.o
 OBJS		+= gpio.o gpio_common_all.o gpio_common_f0234.o
+OBJS		+= exti_common_all.o
```
Comment (karlp) [inline lib/stm32/f7/Makefile]: you don't need it twice...

File: include/libopencmsis/core_cm3.h [hunk 4]
```diff
@@ -26,7 +26,7 @@
  * libopencm3 ones, "volatile" is all that's missing. */
 #define __IO volatile
 #define __O volatile
-#define __I volatile
+#define __I volatile const
```
Comment (karlp) [inline include/libopencmsis/core_cm3.h]: this seems very out of scope for this commit! (and needs further explanation)

File: include/libopencm3/stm32/fsmc.h [hunk 5]
```diff
@@ -23,16 +23,18 @@
...
-
-#define FSMC_BANK1_BASE			0x60000000U /* NOR / PSRAM */
-#define FSMC_BANK2_BASE			0x70000000U /* NAND flash */
-#define FSMC_BANK3_BASE			0x80000000U /* NAND flash */
-#define FSMC_BANK4_BASE			0x90000000U /* PC card */
+// Those macros are not constant along mcus!
+//#define FSMC_BANK1_BASE			0x60000000U /* NOR / PSRAM */
+//#define FSMC_BANK2_BASE			0x70000000U /* NAND flash */
+//#define FSMC_BANK3_BASE			0x80000000U /* NAND flash */
+//#define FSMC_BANK4_BASE			0x90000000U /* PC card */
```
Comment (karlp) [inline include/libopencm3/stm32/fsmc.h]: we don't commit commented out code like this please.

File: include/libopencmsis/core_cm3.h [hunk 6]
```diff
@@ -38,14 +38,27 @@
...
+	__IO uint32_t MMFAR;
+	__IO uint32_t BFAR;
+	__IO uint32_t AFSR;
+	__I  uint32_t PFR[2];
+	__I  uint32_t DFR;
+	__I  uint32_t ADR;
+	__I  uint32_t MMFR[4];
+	__I  uint32_t ISAR[5];
+		 uint32_t RESERVED0[5];
+	__IO uint32_t CPACR;
```
Comment (karlp) [inline include/libopencmsis/core_cm3.h]: aall of this is out of scope for this commit

File: include/libopencm3/stm32/fsmc.h (hunk 5 above)
Comment (karlp) [reply include/libopencm3/stm32/fsmc.h]: Also, you've commented them out, but I dont' see where you've put them back again for f4 and f7?

File: include/libopencm3/stm32/f7/fmc.h [hunk 7]
```diff
@@ -0,0 +1,256 @@
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * Copyright (C) 2013 Chuck McManis <cmcmanis@mcmanis.com>
```
Comment (karlp) [inline include/libopencm3/stm32/f7/fmc.h]: This entire file appears to be a duplicate of the f4 file?  I couldn't spot any differences, please make it a common file too.

File: include/libopencm3/stm32/f7/memorymap.h [hunk 8]
```diff
@@ -93,7 +93,7 @@
 #define SPI6_BASE			(PERIPH_BASE_APB2 + 0x5400)
 #define SAI1_BASE			(PERIPH_BASE_APB2 + 0x5800)
 #define SAI2_BASE			(PERIPH_BASE_APB2 + 0x5C00)
-#define LCD_TFT_BASE			(PERIPH_BASE_APB2 + 0x6800)
+#define LTDC_BASE			(PERIPH_BASE_APB2 + 0x6800)
```
Comment (karlp) [inline include/libopencm3/stm32/f7/memorymap.h]: it's still called LCD-TFT in the ref man, I'm not really keen on changing this, at least not in the same commit.  I'd be more open to just having a second one perhaps?

File: lib/stm32/f7/Makefile [hunk 9]
```diff
@@ -26,7 +26,7 @@ PREFIX		?= arm-none-eabi
 CC		= $(PREFIX)-gcc
 AR		= $(PREFIX)-ar
 
-# STM32F7 only supports single precision FPU
+# Some STM32F7 only support single precision FPU
```
Comment (karlp) [inline lib/stm32/f7/Makefile]: out of scope of this commit, but yes, definitely true :)

File: include/libopencm3/stm32/f7/dsi.h [hunk 10]
```diff
@@ -0,0 +1,25 @@
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * Copyright (C) 2016 Chuck McManis <cmcmanis@mcmanis.com>
```
Comment (karlp) [inline include/libopencm3/stm32/f7/dsi.h]: as elsewhere, I don't feel this is a correct copyright statement for a single include in a new file.

File: include/libopencm3/stm32/f7/rcc.h [hunk 11]
```diff
@@ -571,6 +571,7 @@
 
 /* --- RCC_DCKCFGR2 values -------------------------------------------------- */
 
+#define RCC_DCKCFGR2_DSISEL			(1<<28)
```
Comment (karlp) [inline include/libopencm3/stm32/f7/rcc.h]: what f7 part are you looking at here? I can't find that in the ref mans I'm skimming.

Comment (karlp) [reply include/libopencm3/stm32/f7/rcc.h]: RM0410 has DSISEL on bit 30 at least?

File: include/libopencm3/stm32/f4/rcc.h [hunk 12]
```diff
@@ -681,7 +690,7 @@
 
 /* RCC_PLLSAICFGR[27:24]: PLLSAIQ */
 #define RCC_PLLSAICFGR_PLLSAIQ_SHIFT		24
-#define RCC_PLLSAICFGR_PLLSAIQ_MASK		0xF
+#define RCC_PLLSAICFGR_PLLSAIQ_MASK		0xf
```
Comment (karlp) [inline include/libopencm3/stm32/f4/rcc.h]: pretty gratuitous :) you didn't change other caps to lowercase :)

File: include/libopencm3/stm32/f4/rcc.h [hunk 13]
```diff
@@ -699,8 +708,8 @@
 #define RCC_PLLSAICFGR_PLLSAIN_SHIFT		6
 #define RCC_PLLSAICFGR_PLLSAIN_MASK		0x1FF
 
-
 /* --- RCC_DCKCFGR values -------------------------------------------------- */
+
```
Comment (karlp) [inline include/libopencm3/stm32/f4/rcc.h]: whitespace changes.

File: include/libopencm3/stm32/f4/rcc.h [hunk 14]
```diff
@@ -858,6 +867,7 @@ enum rcc_periph_clken {
 	RCC_WWDG	= _REG_BIT(0x40, 11),
 	RCC_SPI2	= _REG_BIT(0x40, 14),
 	RCC_SPI3	= _REG_BIT(0x40, 15),
+	RCC_SPDIFRX	= _REG_BIT(0x40, 16),
```
Comment (karlp) [inline include/libopencm3/stm32/f4/rcc.h]: You should add in the i2c4, CEC and son on you added above here as well please.

File: include/libopencm3/stm32/f7/rcc.h [hunk 15]
```diff
@@ -518,94 +604,164 @@
...
+/* PLLSAI1 helper macros */
+static inline void rcc_pllsai_enable(void)
+{
+	RCC_CR |= RCC_CR_PLLSAION;
+}
+
+static inline bool rcc_pllsai_ready(void)
+{
+	return (RCC_CR & RCC_CR_PLLSAIRDY) != 0;
+}
```
Comment (karlp) [inline include/libopencm3/stm32/f7/rcc.h]: we've not done this for other things, I don't want to start doing this now.  Just put them in the c files like every other one line function.

File: lib/stm32/common/rcc_common_f47.c [hunk 16]
```diff
@@ -0,0 +1,430 @@
+/** @defgroup rcc_file RCC
+ *
+ * @ingroup STM32F4xx
```
Comment (karlp) [inline lib/stm32/common/rcc_common_f47.c]: this all looks wrong now.

File: lib/stm32/f7/rcc.c [hunk 17]
```diff
@@ -1,11 +1,48 @@
+/** @defgroup rcc_file RCC
+ *
+ * @ingroup STM32F4xx
+ *
+ * @section rcc_f4_api_ex Reset and Clock Control API.
+ *
+ * @brief <b>libopencm3 STM32F4xx Reset and Clock Control</b>
```
Comment (karlp) [inline lib/stm32/f7/rcc.c]: text here is outdated too

File: lib/stm32/f7/rcc.c [hunk 18]
```diff
@@ -1,11 +1,48 @@
...
+ *
+ * LGPL License Terms @ref lgpl_license
+ */
+
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * Copyright (C) 2009 Federico Ruiz-Ugalde <memeruiz at gmail dot com>
+ * Copyright (C) 2009 Uwe Hermann <uwe@hermann-uwe.de>
+ * Copyright (C) 2010 Thomas Otto <tommi@viadmin.org>
```
Comment (karlp) [inline lib/stm32/f7/rcc.c]: I don't really feel comfortable with going and adding copyright statements to an existing file that weren't there before.

File: include/libopencm3/stm32/common/dma2d_common_f47.h [hunk 19]
```diff
@@ -0,0 +1,187 @@
+/** @defgroup dma2d_defines DMA2D Defines
+ *
+ * @ingroup STM32F4xx_defines
+ *
+ * @brief Defined Constants and Macros for the STM32x4xx DMA2D Peripheral
```
Comment (karlp) [inline include/libopencm3/stm32/common/dma2d_common_f47.h]: naming here is off..

File: include/libopencm3/stm32/f7/dma2d.h [hunk 20]
```diff
@@ -0,0 +1,26 @@
+/*
+ * This file is part of the libopencm3 project.
+ *
+ * Copyright (C) 2016 Chuck McManis <cmcmanis@mcmanis.com>
```
Comment (karlp) [inline include/libopencm3/stm32/f7/dma2d.h]: same as elsewhere.

File: include/libopencm3/stm32/common/dma2d_common_f47.h [hunk 21]
```diff
@@ -145,7 +145,8 @@
 #define DMA2D_OPFCCR_CM_ARGB8888	0
 #define DMA2D_OPFCCR_CM_RGB888		1
 #define DMA2D_OPFCCR_CM_RGB565		2
-#define DMA2D_OPFCCR_CM_ARGB4444	3
+#define DMA2D_OPFCCR_CM_ARGB1555	3
+#define DMA2D_OPFCCR_CM_ARGB4444	4
```
Comment (karlp) [inline include/libopencm3/stm32/common/dma2d_common_f47.h]: this one is perfect, I can pick that right now on the original f4 file if you like?

File: include/libopencm3/stm32/common/dsi_common_f47.h [hunk 22]
```diff
@@ -39,779 +39,1528 @@
...
-#define DSI_WRPCR_NDIV_SHIFT		2
-#define DSI_WRPCR_NDIV_MASK		0x7f
-#define DSI_WRPCR_PLLEN			(1 << 0)
+/** @defgroup dsi_registers DSI Registers
+ * @ingroup dsi_defines
+ * @brief DSI Host Control Registers
+@{*/
+
+/* --- DSI Host --- */
+/* DSI_VR:      DSI Host version */
```
Comment (karlp) [inline include/libopencm3/stm32/common/dsi_common_f47.h]: you've made it less lines, and i see you've tried to get the doxygen groupings, but you've dropped the second * in all your comments, so you've actually lost all the documentation that was existing, even if it was verbose.  

/** DSI Host version */ is fine, you can put the define in again if you feel like it, I find it a bit redundant personally, but the second * is critical.

**Comment by karlp** [comment]
for the dsi header, you say, "extend/reorganise dsi header (mostly backwards compatible, except for the typos)"  you've done a lot of doxygen changes by the looks and reformattings?  If there's typos, it's nice ot put them in separately, so they don't get lost in a sea of reformatting.

File: include/libopencm3/stm32/f7/rcc.h (hunk 15 above)
Comment (H2OBrain) [reply include/libopencm3/stm32/f7/rcc.h]: this is from the f4 rcc.h

File: include/libopencm3/stm32/f7/memorymap.h (hunk 8 above)
Comment (H2OBrain) [reply include/libopencm3/stm32/f7/memorymap.h]: it could also be changed in the f4 memorymap

File: lib/stm32/f7/Makefile (hunk 9 above)
Comment (H2OBrain) [reply lib/stm32/f7/Makefile]: ah crap.. yeah, i do those fly-by changes when i am tired

File: include/libopencm3/stm32/f7/exti.h (hunk 2 above)
Comment (H2OBrain) [reply include/libopencm3/stm32/f7/exti.h]: this is from the f4 header file

File: include/libopencmsis/core_cm3.h (hunk 4 above)
Comment (H2OBrain) [reply include/libopencmsis/core_cm3.h]: this seems to be an oversight..

File: include/libopencmsis/core_cm3.h (hunk 6 above)
Comment (H2OBrain) [reply include/libopencmsis/core_cm3.h]: i will just undo all the changes in the core_cm3.h then.
i played a bit with debug stuff and wanted to look at the working register, which should be defined in this struct..

File: include/libopencm3/stm32/f7/fmc.h (hunk 7 above)
Comment (H2OBrain) [reply include/libopencm3/stm32/f7/fmc.h]: the addresses differ, see the f4-to-f7 migration application note an4660, chapter 2.5 (http://www.st.com/content/ccc/resource/technical/document/application_note/73/76/21/47/ef/d0/4c/16/DM00164538.pdf/files/DM00164538.pdf/jcr:content/translations/en.DM00164538.pdf)

File: include/libopencm3/stm32/f7/rcc.h (hunk 11 above)
Comment (H2OBrain) [reply include/libopencm3/stm32/f7/rcc.h]: sorry, this is a bug which is fixed in just the next commit (stm32f4-7: merge/synchronize rcc between f4 and f7)

File: include/libopencm3/stm32/f4/rcc.h (hunk 12 above)
Comment (H2OBrain) [reply include/libopencm3/stm32/f4/rcc.h]: yeah, meld complained :)

File: include/libopencm3/stm32/f4/rcc.h (hunk 13 above)
Comment (H2OBrain) [reply include/libopencm3/stm32/f4/rcc.h]: also meld.. i tried to make f4 and f7 as equal as possible to spot real differences

File: include/libopencm3/stm32/common/dma2d_common_f47.h (hunk 21 above)
Comment (H2OBrain) [reply include/libopencm3/stm32/common/dma2d_common_f47.h]: i don't need it in the f4 file, but go ahead

File: include/libopencm3/stm32/common/dsi_common_f47.h (hunk 22 above)
Comment (H2OBrain) [reply include/libopencm3/stm32/common/dsi_common_f47.h]: this style of commenting i took from the f4 rcc.h file.. I thought it is the standard way of commenting in libopencm3. It would be easy to change in bulk. Do you have a demo header whith comments somewhere?

File: include/libopencm3/stm32/f7/dma2d.h (hunk 20 above)
Comment (H2OBrain) [reply include/libopencm3/stm32/f7/dma2d.h]: should i just remove the whole header then? or just the copyright?

File: include/libopencm3/stm32/f7/fmc.h (hunk 7 above)
Comment (H2OBrain) [reply include/libopencm3/stm32/f7/fmc.h]: but apparently i messed this up somehow

File: include/libopencm3/stm32/f7/rcc.h (hunk 11 above)
Comment (H2OBrain) [reply include/libopencm3/stm32/f7/rcc.h]: ah crap! this is a bug which is fixed in just the next commit

**Comment by H2OBrain** [comment]
I think most of the problems are now fixed. I did not changed the dsi-header documentation style though..

File: include/libopencm3/stm32/common/dsi_common_f47.h (hunk 22 above)
Comment (karlp) [reply include/libopencm3/stm32/common/dsi_common_f47.h]: well, the f4 rcc defines don't show up in documentation either, so they're not a good example :)  With your changes, run "make doc" and look at what comes out.  (You don't get _any _ DSI docs after this change)

**Comment by H2OBrain** [comment]
okay..I broke all the documentation.. this will take a couple of days..

**Comment by H2OBrain** [comment]
> header, you say, "extend/reorganise dsi header (mostly backwards compatible, except for the typos)" you've

i split the changes into two commits now

File: include/libopencm3/stm32/common/dsi_common_f47.h (hunk 22 above)
Comment (H2OBrain) [reply include/libopencm3/stm32/common/dsi_common_f47.h]: The documentation should be fine now

File: lib/stm32/f7/rcc.c (hunk 18 above)
Comment (H2OBrain) [reply lib/stm32/f7/rcc.c]: okay, I removed the copyright statements

File: include/libopencm3/stm32/fsmc.h (hunk 5 above)
Comment (H2OBrain) [reply include/libopencm3/stm32/fsmc.h]: the f4 and f7 headers are now combined

**Comment by H2OBrain** [comment]
okay, i think its complete now.. should i start a new pr?

File: include/libopencm3/stm32/f7/exti.h [hunk 23]
```diff
@@ -0,0 +1,40 @@
+/** @defgroup exti_defines EXTI Defines
+ *
+ * @brief <b>Defined Constants and Types for the STM32F7xx External Interrupts
+ * </b>
+ *
+ * @ingroup STM32F7xx_defines
+ *
+ * @author @htmlonly &copy; @endhtmlonly 2013
+ * Piotr Esden-Tempski <piotr@esden.net>
```
Comment (karlp) [inline include/libopencm3/stm32/f7/exti.h]: I stiull have the same comments on this.  (and the date below)

File: lib/stm32/common/exti_common_all.c [hunk 24]
```diff
@@ -17,7 +17,7 @@
  * You should have received a copy of the GNU Lesser General Public License
  * along with this library.  If not, see <http://www.gnu.org/licenses/>.
  *
- * This provides the code for the "next gen" EXTI block provided in F2/F4/L1
+ * This provides the code for the "next gen" EXTI block provided in F2/F4/F7/L1
```
Comment (karlp) [inline lib/stm32/common/exti_common_all.c]: If you want to touch this line, you'd be safer saying, "non-f1" rather than leaving it still incomplete. (l0,l4, f3, f0...)

though, in _reality_ this file is actually included for f1 too, so this was just badly copied from somewhere else in the past.

File: include/libopencm3/stm32/fsmc.h [hunk 25]
```diff
@@ -23,16 +23,17 @@
 #include <libopencm3/cm3/common.h>
 #include <libopencm3/stm32/memorymap.h>
 
-#if defined(STM32F4)
+#if   defined(STM32F4)
```
Comment (karlp) [inline include/libopencm3/stm32/fsmc.h]: whitespace change

File: include/libopencm3/stm32/fsmc.h [hunk 26]
```diff
@@ -23,16 +23,17 @@
...
 #endif
 
 /* --- Convenience macros -------------------------------------------------- */
-
 #define FSMC_BANK1_BASE			0x60000000U /* NOR / PSRAM */
-#define FSMC_BANK2_BASE			0x70000000U /* NAND flash */
+#define FSMC_BANK2_BASE			0x70000000U /* NAND flash (reserved in F7) */
 #define FSMC_BANK3_BASE			0x80000000U /* NAND flash */
-#define FSMC_BANK4_BASE			0x90000000U /* PC card */
+#define FSMC_BANK4_BASE			0x90000000U /* PC card (reserved in F7) */
```
Comment (karlp) [inline include/libopencm3/stm32/fsmc.h]: it's not reserved, it's quadspi.  If these are not "common" then they should simply be moved into the f4 or f7 files explicitly.

File: include/libopencm3/stm32/ltdc.h [hunk 27]
```diff
@@ -22,6 +22,8 @@
 
 #if defined(STM32F4)
 #       include <libopencm3/stm32/f4/ltdc.h>
+#elif defined(STM32F7)
+#       include <libopencm3/stm32/f7/ltdc.h>
 #else
 #       error "LCD-TFT only defined for STM32F4"
```
Comment (karlp) [inline include/libopencm3/stm32/ltdc.h]: you could make this line generic too, but no stress

File: include/libopencm3/stm32/common/dsi_common_f47.h [hunk 28]
```diff
@@ -44,779 +44,1667 @@
...
+	DSI_CMCR_GLWTX                   = 1<<14,
+	DSI_CMCR_GSR2TX                  = 1<<13,
+	DSI_CMCR_GSR1TX                  = 1<<12,
+	DSI_CMCR_GSR0TX                  = 1<<11,
+	DSI_CMCR_GSW2TX                  = 1<<10,
+	DSI_CMCR_GSW1TX                  = 1<<9,
+	DSI_CMCR_GSW0TX                  = 1<<8,
+	DSI_CMCR_ARE                     = 1<<1,
+	DSI_CMCR_TEARE                   = 1<<0
+} dsi_cmcr_flags_t;
```
Comment (karlp) [inline include/libopencm3/stm32/common/dsi_common_f47.h]: I'm not _at all_ a fan of trying this alternative style of defining all the fields. it's not done anywhere else, and it adds lots of typedefs.  What's the motivation for going for a completely new style for bit definitions?

File: include/libopencmsis/core_cm3.h [hunk 29]
```diff
@@ -36,16 +36,35 @@
 
 /* structure as in, for example,
  * DeviceSupport/EnergyMicro/EFM32/efm32tg840f32.h, data from
- * libopencm3/cm3/scb.h. FIXME incomplete. */
+ * libopencm3/cm3/scb.h */
 typedef struct {
-	__IO uint32_t CPUID;
+	__I  uint32_t CPUID;
```
Comment (karlp) [inline include/libopencmsis/core_cm3.h]: This commit needs, as mentioned, much more information abotu what you're doing.  Good to see it in a separate commit, but it's still completely without context.

File: include/libopencm3/stm32/f7/rcc.h (hunk 11 above)
Comment (karlp) [reply include/libopencm3/stm32/f7/rcc.h]: please don't add it in wrongly in this one at all.

**Comment by karlp** [comment]
lots of this has now been merged. the remaining will need to be rebased and go into a new pr for further discussion,  thanks for this!
